File: pbark/pbark/parser/keywords/attribute_keyword_groups.py
from __future__ import annotations

from pbark.parser.keywords.keyword_registry import KeywordRegistry


class AttributeKeywordGroups:
    def __init__(self, registry: KeywordRegistry) -> None:
        self.trait_words = registry.heard_words("loud", "greedy", "lazy", "wet", "fetchy", "chaser", "playful")
        self.food_increment_words = registry.heard_words("feed", "feeds", "refills", "steals", "pinch", "pinches")
        self.item_increment_words = registry.heard_words(
            "grow", "grows", "grew", "gets", "receives", "earns", "gains", "collects", "snags", "nabs",
            "inherits", "scores", "find", "finds", "fetch", "fetches", "discovers", "unearths",
        )
        self.increment_words = list(dict.fromkeys(self.food_increment_words + self.item_increment_words))
        self.food_decrement_words = registry.heard_words(
            "gulp", "gulps", "eat", "eats", "ate", "devour", "devours", "chews", "munches", "scarfs", "wolfs", "nibbles"
        )
        self.item_decrement_words = registry.heard_words(
            "buried", "lost", "misplaces", "breaks", "destroys", "hides", "buries", "forgets", "tosses", "abandons"
        )
        self.decrement_words = list(dict.fromkeys(self.food_decrement_words + self.item_decrement_words))
        self.type_number_words = registry.heard_words("number", "digit", "digits")
        self.type_words_words = registry.heard_words("words", "text")
        self.type_nothing_words = registry.heard_words("nothing")
        self.sniff_words = registry.heard_words("sniffs", "smells", "scents", "checks", "peeks")
        self.growl_words = registry.heard_words("growls", "roars", "bellows")
        self.same_words = registry.phrase_words("same")
        self.than_words = registry.phrase_words("than")
        self.at_words = registry.phrase_words("at")
        self.list_and_words = registry.phrase_words("and")
        self.the_words = registry.phrase_words("the")
        self.of_words = registry.phrase_words("of")
        self.items_words = registry.phrase_words("items")
        self.in_words = registry.phrase_words("in", "into")
        self.as_words = registry.phrase_words("as")
        self.by_words = registry.phrase_words("by")
        self.how_words = registry.phrase_words("how")
        self.many_words = registry.phrase_words("many")
        self.years_words = registry.phrase_words("years", "year")
        self.old_words = registry.phrase_words("old")
        self.together_words = registry.phrase_words("together")
        self.or_words = registry.phrase_words("or")

File: src/main/java/dev/klomptech/jbark/parser/keywords/AttributeKeywordGroups.java
package dev.klomptech.jbark.parser.keywords;

import java.util.ArrayList;
import java.util.List;

public final class AttributeKeywordGroups {

  public final List<String> traitWords;
  public final List<String> incrementWords;
  public final List<String> foodIncrementWords;
  public final List<String> itemIncrementWords;
  public final List<String> decrementWords;
  public final List<String> foodDecrementWords;
  public final List<String> itemDecrementWords;
  public final List<String> typeNumberWords;
  public final List<String> typeWordsWords;
  public final List<String> typeNothingWords;
  public final List<String> sniffWords;
  public final List<String> growlWords;
  public final List<String> sameWords;
  public final List<String> thanWords;
  public final List<String> atWords;
  public final List<String> listAndWords;
  public final List<String> theWords;
  public final List<String> ofWords;
  public final List<String> itemsWords;
  public final List<String> inWords;
  public final List<String> asWords;
  public final List<String> byWords;
  public final List<String> howWords;
  public final List<String> manyWords;
  public final List<String> yearsWords;
  public final List<String> oldWords;
  public final List<String> togetherWords;
  public final List<String> orWords;

  public AttributeKeywordGroups(final KeywordRegistry registry) {
    traitWords =
        registry.heardWords("loud", "greedy", "lazy", "wet", "fetchy", "chaser", "playful");
    foodIncrementWords =
        registry.heardWords("feed", "feeds", "refills", "steals", "pinch", "pinches");
    itemIncrementWords =
        registry.heardWords(
            "grow",
            "grows",
            "grew",
            "gets",
            "receives",
            "earns",
            "gains",
            "collects",
            "snags",
            "nabs",
            "inherits",
            "scores",
            "find",
            "finds",
            "fetch",
            "fetches",
            "discovers",
            "unearths");
    incrementWords = concat(foodIncrementWords, itemIncrementWords);
    foodDecrementWords =
        registry.heardWords(
            "gulp", "gulps", "eat", "eats", "ate", "devour", "devours", "chews", "munches",
            "scarfs", "wolfs", "nibbles");
    itemDecrementWords =
        registry.heardWords(
            "buried",
            "lost",
            "misplaces",
            "breaks",
            "destroys",
            "hides",
            "buries",
            "forgets",
            "tosses",
            "abandons");
    decrementWords = concat(foodDecrementWords, itemDecrementWords);
    typeNumberWords = registry.heardWords("number", "digit", "digits");
    typeWordsWords = registry.heardWords("words", "text");
    typeNothingWords = registry.heardWords("nothing");
    sniffWords = registry.heardWords("sniffs", "smells", "scents", "checks", "peeks");
    growlWords = registry.heardWords("growls", "roars", "bellows");
    sameWords = registry.phraseWords("same");
    thanWords = registry.phraseWords("than");
    atWords = registry.phraseWords("at");
    listAndWords = registry.phraseWords("and");
    theWords = registry.phraseWords("the");
    ofWords = registry.phraseWords("of");
    itemsWords = registry.phraseWords("items");
    inWords = registry.phraseWords("in", "into");
    asWords = registry.phraseWords("as");
    byWords = registry.phraseWords("by");
    howWords = registry.phraseWords("how");
    manyWords = registry.phraseWords("many");
    yearsWords = registry.phraseWords("years", "year");
    oldWords = registry.phraseWords("old");
    togetherWords = registry.phraseWords("together");
    orWords = registry.phraseWords("or");
  }

  private static List<String> concat(final List<String> first, final List<String> second) {
    List<String> merged = new ArrayList<>(first);
    merged.addAll(second);
    return List.copyOf(merged);
  }
}
